std/container/rbtree.d

```diff
@@ -1094,6 +1094,167 @@ if (is(typeof(binaryFun!less(T.init, T.init))))
         assert(t1 != o);  // pathological case, must not crash
     }
 
+    /**
+     * Generates a hash for the tree. Note that with a custom comparison function
+     * it may not hold that if two rbtrees are equal, the hashes of the trees
+     * will be equal.
+     */
+    override size_t toHash() nothrow @safe
+    {
+        import std.traits : hasMember;
+
+        auto thisRange = this[];
+
+        static if (hasMember!(Elem, "toHash"))
+        {
+            size_t hash = 0;
+
+            static if (is(T == class) || is(T == interface))
+            {
+                foreach (ref e; thisRange)
+                {
+                    hash += (e is null ? 0 : e.toHash()) + 0x9e3779b9 + (hash << 6) + (hash >>> 2);
```

Comment:
The hash number could have been at least a magic constant, though repeating the hashing code three times isn't really great either. I would have gone with an inlineable function or mixin.